[MDS-6205] Permit condition specific reports

matbusby-fw opened the pull request:
## Objective 

- Created a mine_report_permit_requirement table to set up permit required reports with frequency that can be attached to a permit condition.
- Updated front and backend to interface with this new table in the permit conditions view
- Set up create of a mine_report_permit_requirement flow

[MDS-6205](https://bcmines.atlassian.net/browse/MDS-6205)

![permit-reports](https://github.com/user-attachments/assets/a7759357-54f2-4799-be28-fbbdf7536a66)

matbusby-fw committed:
addressed code quality concern from sonarcloud

update PermitConditions.spec.tsx.snap

update conditions to fix propogation issue

update RenderRadioButtons.spec.js.snap

Add report prop to permit condition form and disable button if report exists

This commit introduces a new optional `report` prop to the `AddReportToPermitConditionForm` component, allowing it to use an existing report if available. Additionally, it modifies the permit condition layer to disable the "Add Report" button when a report already exists for the sub-condition, and ensures permits are refetched after adding a new report.

Rename component and update references

Renamed `AddRequestToPermitConditionForm` to `AddReportToPermitConditionForm` to better reflect its purpose. Also updated all references and added a default export for consistency. Added a snapshot test for the form component.

Add report handling to permit conditions

Enhanced permit conditions to include associated reports. Updated the UI to display report-related information and modified relevant constants, styles, and selectors accordingly.

update interfaces to accomodate permit reports

added permit condition specific report form

updated action buttons on specific permit conditions and updated interface to expand and show action buttons when clicking into a condition.

update backend to accomodate new permit condition specific details for mine reports

migrate mine_report table with new permit specific details

matbusby-fw committed:
revert backend changes

matbusby-fw committed:
create new mine_report_permit_requirement.py table and updated permit return to include these.

matbusby-fw committed:
Rename form and update permit requirement logic

Renamed `AddReporttoPermitConditionForm` to `ReportPermitRequirementForm`. Updated the form to use new `mineReportPermitRequirement` structure. Enhanced `MineReportPermitRequirement` model with soft delete and additional fields like `initial_due_date` and `ministry_recipient`.

matbusby-fw committed:
added ts-ignore

matbusby-fw committed:
updated text for adding a report requirement

matbusby-fw committed:
make view mode RenderDate formatting conditional

taraepp commented on services/core-web/src/components/Forms/reports/ReportPermitRequirementForm.tsx:
```diff
@@ -0,0 +1,195 @@
+import RenderCancelButton from "@mds/common/components/forms/RenderCancelButton";
+import RenderField from "@mds/common/components/forms/RenderField";
+import RenderRadioButtons from "@/components/common/RenderRadioButtons";
+import { requiredRadioButton } from "@common/utils/Validate";
```
would want to update the import for requiredRadioButton and RenderRadioButtons to be from @mds/common

taraepp commented on services/core-web/src/components/common/RenderRadioButtons.tsx:
```diff
@@ -46,8 +47,17 @@ const RenderRadioButtons: FC<RenderRadioButtonsProps> = (props) => {
         id={id as string}
-        options={options}
-      />
+      >
```
the common version has isVertical instead of vertical for a parameter. Just something to keep in mind

taraepp commented on services/core-web/src/components/mine/Permit/PermitConditionLayer.tsx:
```diff
@@ -1,32 +1,185 @@
+        {sectionEdit && (
+          <Row justify="space-between">
+            <Row gutter={8} className="condition-edit-buttons" align="middle">
+              <Button
```
I think that all the children of `<Row>` need to be nested under a `<Col>` element(s). It may work a little better/as expected and make some of the added CSS unnecessary.

matbusby-fw replied:
Ah, yep.  Just missed this one.

taraepp commented on services/core-web/src/components/mine/Permit/PermitConditions.tsx:
```diff
@@ -65,14 +74,43 @@ const PermitConditions: FC<PermitConditionProps> = ({
+      };
+
+      // Initialize the step paths for all top-level conditions
+      const updatedConditions = conditions.map((condition) => getStepPath(condition));
```
perhaps a bit of a nitpick, but are these _updated_ conditions, as in they have updates that the user is making that need to be saved? Or are they _formatted_?

matbusby-fw replied:
Fair enough.  I'll update it!

taraepp commented:
I am still a bit unsure if what the condition requirement should be mapped to is the permit or the permit amendment (granted, the permit condition is _already_ mapped to the permit amendment which is mapped to the permit but I get trying to avoid excessively complicated queries). But I also haven't tried running or rendering it. The decider to me would be "What happens when a permit has 2 amendments, and each of them have conditions with different report requirements?" <-- if it can handle that as is, then I'm happy. May just need to have dataMocks be a little more fleshed out with such a scenario.

matbusby-fw committed:
address PR comments and added backend resource test.

matbusby-fw committed:
renamed migration and added cols to buttons on permit conditions

matbusby-fw commented:
> I am still a bit unsure if what the condition requirement should be mapped to is the permit or the permit amendment (granted, the permit condition is _already_ mapped to the permit amendment which is mapped to the permit but I get trying to avoid excessively complicated queries). But I also haven't tried running or rendering it. The decider to me would be "What happens when a permit has 2 amendments, and each of them have conditions with different report requirements?" <-- if it can handle that as is, then I'm happy. May just need to have dataMocks be a little more fleshed out with such a scenario.

It worked since it was just using it to grab the requirements for that permit, and since the important connection was between permit_condition and the requirement it would maintain the correct usage.  That said, I think you're right that amendment is actually more appropriate, so I updated it to adhere to that.

matbusby-fw committed:
update snap

matbusby-fw committed:
update permitAmendment factory to create auth end date as date instead of datetime

matbusby-fw committed:
update mine_report_permit_requirement response model to not break swagger

taraepp commented on services/common/src/redux/selectors/permitSelectors.ts:
```diff
@@ -84,3 +85,18 @@ export const getPermits = createSelector([getUnformattedPermits], (permits) => {
+
+export const getMineReportPermitRequirements = (permitGuid) =>
+  createSelector(
+    [getPermitByGuid(permitGuid)],
```
minor- but I would recommend replacing getPermitByGuid with getLatestAmendmentByPermitGuid here.

taraepp commented on services/core-api/app/api/mines/reports/models/mine_report_permit_requirement.py:
```diff
@@ -0,0 +1,87 @@
+    ministry_recipient: Optional[list[OfficeDestination]] = db.Column(
+        ARRAY(db.Enum(OfficeDestination, name='ministry_recipient_type')), nullable=True)
+    permit_condition_id: int = db.Column(db.Integer, db.ForeignKey('permit_conditions.permit_condition_id'))
+    permit_amendment_id: int = db.Column(db.Integer, db.ForeignKey('permit_amendment.permit_amendment_id'))
```
Wow, so much typing! 🎨

taraepp commented on services/core-web/src/components/mine/Permit/PermitConditionLayer.tsx:
```diff
@@ -1,32 +1,206 @@
+                  <Button
+                    onClick={(event) => {
+                      event.stopPropagation();
+                      console.log("Not yet implemented");
```
Just curious what the feature flag status of this is. I'm seeing a Feature.MODIFY_PERMIT_CONDITIONS (separate from Feature.PERMIT_CONDITIONS_PAGE) that should possibly be blocking these buttons.

matbusby-fw replied:
Isn't this whole page behind a feature flag at this point in time?  I should check that.

taraepp replied:
Technically the page is behind PERMIT_CONDITIONS_PAGE (which is currently off) but we have modifications controlled by the MODIFY flag separately. So we could make it possible to _view_ data and not _change_ anything. I think it would be nice to be able to hide the unimplemented features, specifically.

taraepp commented on services/core-web/src/components/mine/Permit/PermitConditions.tsx:
```diff
@@ -184,7 +270,7 @@ const PermitConditions: FC<PermitConditionProps> = ({
           </Col>
         </Row>
       }
-    />
```
I think this one wants to stay a self-closing tag? It doesn't take any children

matbusby-fw replied:
That's weird.  I made no concious decision to change that.

simensma-fresh commented on services/core-api/app/api/mines/reports/resources/mine_report_permit_requirement.py:
```diff
@@ -0,0 +1,74 @@
+from app.api.utils.resources_mixins import UserMixin
+
+
+class MineReportPermitRequirementResource(Resource, UserMixin):
```
Would be great if we could add an entry for this endpoint in  `test_expected_auth.py`

matbusby-fw replied:
👍🏻

simensma-fresh commented on services/core-api/app/api/utils/helpers.py:
```diff
@@ -2,6 +2,9 @@
 from datetime import datetime
+
+from flask_restx import ValidationError
+from flask_restx.fields import Raw
```
Intentional to add these?

matbusby-fw replied:
Ah, I removed the thing that used them.  I'll remove the imports

matbusby-fw committed:
Enhance permit condition editing permissions

Added feature flag check to control permit condition editing. Updated PermitConditionLayer component to respect user permissions and feature availability. Simplified getMineReportPermitRequirements selector for clarity.

matbusby-fw committed:
Remove unused imports and add new test case for MineReportPermit

Removed `ValidationError` and `Raw` imports from `helpers.py` as they were not being used. Added a test case in `test_expected_auth.py` for `MineReportPermitRequirementResource` to ensure proper resource handling and permissions.
